src/olympia/devhub/views.py

```diff
@@ -1387,109 +1390,207 @@ def version_stats(request, addon_id, addon):
+        else:
+            addon = Addon.from_upload(
+                upload=data['upload'],
+                platforms=data.get('supported_platforms', []),
+                source=data['source'],
+                is_listed=channel == amo.RELEASE_CHANNEL_LISTED)
+            version = addon.current_version
             AddonUser(addon=addon, user=request.user).save()
-            check_validation_override(request, form, addon,
-                                      addon.current_version)
-            if not addon.is_listed:  # Not listed? Automatically choose queue.
-                addon.update(status=amo.STATUS_NOMINATED)
-                # Sign all the files submitted, one for each platform.
-                auto_sign_version(addon.versions.get())
-                return redirect('devhub.submit.finish', addon.slug)
-            return redirect('devhub.submit.details', addon.slug)
-    is_admin = acl.action_allowed(request, 'ReviewerAdminTools', 'View')
+            url_args = [addon.slug]
 
+        check_validation_override(request, form, addon, version)
+        if data['source']:
+            addon_update = {'admin_review': True}
+        else:
+            addon_update = {}
+        if addon.status == amo.STATUS_NULL and addon.has_complete_metadata():
+            addon_update.update(status=amo.STATUS_NOMINATED)
+        if addon_update:
+            addon.update(**addon_update)
+        # auto-sign versions (the method checks eligibility)
+        auto_sign_version(version, is_beta=is_beta)
+        next_url = (next_listed if channel == amo.RELEASE_CHANNEL_LISTED
+                    else next_unlisted)
+        return redirect(next_url, *url_args)
+    is_admin = acl.action_allowed(request, 'ReviewerAdminTools', 'View')
+    if addon:
+        channel_choice_text = (forms.DistributionChoiceForm().LISTED_LABEL
+                               if channel == amo.RELEASE_CHANNEL_LISTED else
+                               forms.DistributionChoiceForm().UNLISTED_LABEL)
+    else:
+        channel_choice_text = ''  # We only need this for Version upload.
     return render(request, 'devhub/addons/submit/upload.html',
                   {'new_addon_form': form, 'is_admin': is_admin,
-                   'listed': is_listed})
+                   'addon': addon, 'submit': 'version' if addon else 'addon',
+                   'listed': channel == amo.RELEASE_CHANNEL_LISTED,
+                   'channel_choice_text': channel_choice_text})
 
 
-@dev_required(submitting=True)
-def submit_details(request, addon_id, addon):
-    forms_, context = [], {}
+@login_required
+def submit_addon_upload(request, channel):
+    channel_id = amo.CHANNEL_CHOICES_LOOKUP[channel]
+    return _submit_upload(request, None, channel_id,
+                          'devhub.submit.details', 'devhub.submit.finish')
+
+
+@dev_required
+@waffle_switch('step-version-upload')
+def submit_version_upload(request, addon_id, addon, channel):
```

Comment:
really the only difference (now) between this as `submit_version` is the url pattern needs a channel param:
`^versions/submit/upload-(?P<channel>listed|unlisted)$` vs. `^versions/submit/$`.  I'm undecided on ideal urls.